spec/abi.dd

```diff
@@ -957,32 +957,70 @@ $(H2 $(LNAME2 garbage_collection, Garbage Collection))
 
         $(P The interface to this is found in Druntime's $(DRUNTIMESRC core/gc/gcinterface.d).)
 
-$(H2 $(LNAME2 runtime_helper_functions, Runtime Helper Functions))
+$(H2 $(LNAME2 ModuleInfo, ModuleInfo Instance))
 
-        $(P These are found in Druntime's $(DRUNTIMESRC rt/).)
+        $(P An instance of $(LINK2 https://dlang.org/phobos/object.html#.ModuleInfo, `ModuleInfo`)
+        is generated by the compiler and inserted into the object file for every module.
+        `ModuleInfo` contains information about the module that is useful to the D runtime library:
+        )
+
+        $(UL
+        $(LI If the module has a static constructor, static destructor, shared static constructor, or shared static destructor.)
+        $(LI A reference to any unit tests defined by the module.)
+        $(LI An array of references to any imported modules that have one or more of:
+            $(OL
+            $(LI static constructors)
+            $(LI static destructors)
+            $(LI shared static constructors)
+            $(LI shared static destructors)
+            $(LI unit tests)
+            $(LI transitive imports of any module that contains one or more of 1..5)
+            $(LI order independent constructors (currently needed for implementing
+            $(DDSUBLINK dmd, switch-cov, $(TT -cov))))
+            )
+            This enables the runtime to run the unit tests,
+            the module constructors in a depth-first order,
+            and the module destructors in the reverse order.
+        )
+
+        $(LI An array of references to `ClassInfo` for each class defined in the module.
+        $(NOTE this feature may be removed.))
+        )
+
+        $(P `ModuleInfo` is defined in Druntime's $(DRUNTIMESRC object.d).)
+
+        $(P Modules compiled with $(DDSUBLINK dmd, switch-betterC, $(TT -betterC))
+        do not have a `ModuleInfo` instance generated, because such modules must work
+        without the D runtime library.
```

Comment:
A ModuleInfo only contains a reference to another ModuleInfo if that other ModuleInfo has module ctors, dtors, etc.

Reply:
I plan on documenting all this stuff better, as there's a lot of confusion about it. This is the first step.

Reply:
In that case, there's no need for a specific note about betterC, since a betterC module can not have module ctor, etc, nor import anything that does. The one rule ought to apply everywhere.

Reply:
Though betterC doesn't follow that rule right now, it should... cuz right now you can generate a linker error easily by having a middle-man module compiled with -betterC that imports a module with a static ctor.

Reply:
The idea here is to document what is happening, not redesign it. People are clearly confused about what happens with betterC, so it is best to document it.

Reply:
Well, what you wrote in the diff is *not* what is actually happening. If you deleted the " because such modules must work        without the D runtime library" part, it would be better. Generating ModuleInfo will never prevent a module from working without the D runtime library. betterC modules not generating moduleinfo simply breaks transitive dependencies in real D code.